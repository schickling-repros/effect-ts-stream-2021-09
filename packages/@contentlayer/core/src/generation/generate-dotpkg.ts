import * as utils from '@contentlayer/utils'
import type { MkdirError } from '@contentlayer/utils/node'
import { fileOrDirExists } from '@contentlayer/utils/node'
import { pipe } from '@effect-ts/core'
import * as Chunk from '@effect-ts/core/Collections/Immutable/Chunk'
import * as T from '@effect-ts/core/Effect'
import * as S from '@effect-ts/core/Effect/Experimental/Stream'
import * as OT from '@effect-ts/otel'
import { camelCase } from 'camel-case'
import { promises as fs, watch } from 'fs'
import * as path from 'path'
import type { Observable } from 'rxjs'
import { combineLatest, defer, of } from 'rxjs'
import { switchMap, tap } from 'rxjs/operators'
import type { PackageJson } from 'type-fest'

import type { Cache } from '../cache'
import type { PluginOptions, SourcePlugin, SourcePluginType } from '../plugin'
import type { DocumentTypeDef, SchemaDef } from '../schema'
import { makeArtifactsDir, makeArtifactsDirEff } from '../utils'
import { autogeneratedNote } from './common'
import { renderTypes } from './generate-types'

/**
 * Used to track which files already have been written.
 * Gets re-initialized per `generateDotpkg` invocation therefore only "works" during dev mode.
 */
type FilePath = string
type DocumentHash = string
type WrittenFilesCache = Record<FilePath, DocumentHash>

export type GenerationOptions = {
  sourcePluginType: SourcePluginType
  options: PluginOptions
}

// TODO make sure unused old generated files are removed
export const generateDotpkg = ({ source, watchData }: { source: SourcePlugin; watchData: boolean }): Observable<void> =>
  combineLatest({
    cache: source.fetchData({ watch: watchData }),
    schemaDef: defer(async () => source.provideSchema()),
    targetPath: defer(makeArtifactsDir).pipe(
      tap((artifactsDir) => watchData && errorIfArtifactsDirIsDeleted({ artifactsDir })),
    ),
    generationOptions: of({ sourcePluginType: source.type, options: source.options }),
    writtenFilesCache: of({}),
  }).pipe(switchMap(writeFilesForCache))

export const generateDotpkgSingle = ({
  source,
}: {
  source: SourcePlugin
}): T.Effect<OT.HasTracer, Error | MkdirError, void> =>
  pipe(
    generateDotpkgEff({ source, watchData: false }),
    T.tap(() => T.succeedWith(() => console.log('worked'))),
    // S.take(1),
    // S.runCollect,
    // S.runDrain,
    // T.map(Chunk.unsafeHead),
    // OT.withSpan('@contentlayer/core/generation:generateDotpkgSingle', { attributes: {} }),
  )

export const generateDotpkgEff = ({ source, watchData }: { source: SourcePlugin; watchData: boolean }) => {
  // }): S.Stream<OT.HasTracer, Error | MkdirError, void> => {
  // const writtenFilesCache = {}
  // const generationOptions = { sourcePluginType: source.type, options: source.options }
  // const resolveParams = pipe(
  //   T.structPar({
  //     schemaDef: source.provideSchemaEff!,
  //     targetPath: makeArtifactsDirEff,
  //     // __: T.fail(new Error('oh no')),
  //     // .pipe(
  //     //   tap((artifactsDir) => watchData && errorIfArtifactsDirIsDeleted({ artifactsDir }))
  //     // ),
  //   }),
  // )

  const dataStream = source.fetchDataEff!({ watch: true })

  console.log({ dataStream })

  const drainedStream_ = dataStream as unknown as T.Effect<any, never, Cache>

  return pipe(
    drainedStream_,
    // dataStream,
    // S.take(1),
    // S.runDrain,
  )
}

const writeFilesForCacheEff = (params: {
  schemaDef: SchemaDef
  cache: Cache
  targetPath: string
  generationOptions: GenerationOptions
  writtenFilesCache: WrittenFilesCache
}): T.Effect<unknown, Error, void> =>
  T.tryCatchPromise(
    () => writeFilesForCache(params),
    (e) => e as Error,
  )

const writeFilesForCache = (async ({
  cache,
  schemaDef,
  targetPath,
  generationOptions,
  writtenFilesCache,
}: {
  schemaDef: SchemaDef
  cache: Cache
  targetPath: string
  generationOptions: GenerationOptions
  writtenFilesCache: WrittenFilesCache
}): Promise<void> => {
  console.log('write files')

  const withPrefix = (...path_: string[]) => path.join(targetPath, ...path_)

  if (process.env['CL_DEBUG']) {
    // NOTE cache directory already exists because `source.fetchData` has already created it
    await fs.writeFile(withPrefix('cache', 'schema.json'), JSON.stringify(schemaDef, null, 2))
  }

  const allCacheItems = Object.values(cache.cacheItemsMap)
  const allDocuments = allCacheItems.map((_) => _.document)

  const documentDefs = Object.values(schemaDef.documentTypeDefMap)

  const typeNameField = generationOptions.options.fieldOptions.typeFieldName
  const dataBarrelFiles = documentDefs.map((docDef) => ({
    content: makeDataExportFile({
      docDef,
      documentIds: allDocuments.filter((_) => _[typeNameField] === docDef.name).map((_) => _._id),
    }),
    filePath: withPrefix('data', `${getDataVariableName({ docDef })}.js`),
  }))

  const dataJsonFiles = allCacheItems.map(({ document, documentHash }) => ({
    content: JSON.stringify(document, null, 2),
    filePath: withPrefix('data', document[typeNameField], `${idToFileName(document._id)}.json`),
    documentHash,
  }))

  const dataDirPaths = documentDefs.map((_) => withPrefix('data', _.name))
  await Promise.all([mkdir(withPrefix('types')), ...dataDirPaths.map(mkdir)])

  const writeFile = writeFileWithWrittenFilesCache({ writtenFilesCache })

  await Promise.all([
    writeFile({ filePath: withPrefix('package.json'), content: makePackageJson() }),
    writeFile({
      filePath: withPrefix('types', 'index.d.ts'),
      content: renderTypes({ schemaDef, generationOptions }),
    }),
    writeFile({ filePath: withPrefix('types', 'index.js'), content: makeHelperTypes() }),
    writeFile({ filePath: withPrefix('data', 'index.d.ts'), content: makeDataTypes({ schemaDef }) }),
    writeFile({ filePath: withPrefix('data', 'index.js'), content: makeIndexJs({ schemaDef }) }),
    ...dataBarrelFiles.map(writeFile),
    ...dataJsonFiles.map(writeFile),
  ])
})['|>'](
  utils.traceAsyncFn('@contentlayer/core/commands/generate-dotpkg:writeFilesForCache', (_) => utils.omit(_, ['cache'])),
)

const makePackageJson = (): string => {
  const packageJson: PackageJson & { typesVersions: any } = {
    name: 'dot-contentlayer',
    description: 'This package is auto-generated by Contentlayer',
    version: '0.0.0',
    exports: {
      './data': {
        import: './data/index.js',
      },
      './types': {
        import: './types/index.js',
      },
    },
    typesVersions: {
      '*': {
        data: ['./data'],
        types: ['./types'],
      },
    },
  }

  return JSON.stringify(packageJson, null, 2)
}

const mkdir = async (dirPath: string) => {
  try {
    await fs.mkdir(dirPath, { recursive: true })
  } catch (e: any) {
    if (e.code !== 'EEXIST') {
      throw e
    }
  }
}

/**
 * Remembers which files already have been written to disk.
 * If no `documentHash` was provided, the writes won't be cached. */
const writeFileWithWrittenFilesCache =
  ({ writtenFilesCache }: { writtenFilesCache: WrittenFilesCache }) =>
  async ({
    filePath,
    content,
    documentHash,
  }: {
    filePath: string
    content: string
    documentHash?: string
  }): Promise<void> => {
    if (documentHash !== undefined && writtenFilesCache[filePath] === documentHash) {
      return
    }

    await fs.writeFile(filePath, content, 'utf8')
    if (documentHash) {
      writtenFilesCache[filePath] = documentHash
    }
  }

const makeDataExportFile = ({ docDef, documentIds }: { docDef: DocumentTypeDef; documentIds: string[] }): string => {
  const dataVariableName = getDataVariableName({ docDef })

  if (docDef.isSingleton) {
    const documentId = documentIds[0]
    return `\
// ${autogeneratedNote}
export { default as ${dataVariableName} } from './${docDef.name}/${idToFileName(documentId)}.json'
`
  }

  const makeVariableName = utils.flow(idToFileName, (_) => camelCase(_, { stripRegexp: /[^A-Z0-9\_]/gi }))

  const docImports = documentIds
    .map((_) => `import ${makeVariableName(_)} from './${docDef.name}/${idToFileName(_)}.json'`)
    .join('\n')

  return `\
// ${autogeneratedNote}

${docImports}

export const ${dataVariableName} = [${documentIds.map((_) => makeVariableName(_)).join(', ')}]
`
}

const makeIndexJs = ({ schemaDef }: { schemaDef: SchemaDef }): string => {
  const dataVariableNames = Object.values(schemaDef.documentTypeDefMap).map(
    (docDef) => [docDef, getDataVariableName({ docDef })] as const,
  )
  const constReexports = dataVariableNames
    .map(([, dataVariableName]) => `export * from './${dataVariableName}.js'`)
    .join('\n')

  const constImportsForAllDocuments = dataVariableNames
    .map(([, dataVariableName]) => `import { ${dataVariableName} } from './${dataVariableName}.js'`)
    .join('\n')

  const allDocuments = dataVariableNames
    .map(([docDef, dataVariableName]) => (docDef.isSingleton ? dataVariableName : `...${dataVariableName}`))
    .join(', ')

  return `\
// ${autogeneratedNote}

export { isType } from 'contentlayer/client'

${constReexports}
${constImportsForAllDocuments}

export const allDocuments = [${allDocuments}]
`
}

const makeHelperTypes = (): string => {
  return `\
// ${autogeneratedNote}

export { isType } from 'contentlayer/client'
`
}

const makeDataTypes = ({ schemaDef }: { schemaDef: SchemaDef }): string => {
  const dataConsts = Object.values(schemaDef.documentTypeDefMap)
    .map((docDef) => [docDef, docDef.name, getDataVariableName({ docDef })] as const)
    .map(
      ([docDef, typeName, dataVariableName]) =>
        `export declare const ${dataVariableName}: ${typeName}${docDef.isSingleton ? '' : '[]'}`,
    )
    .join('\n')

  const documentTypeNames = Object.values(schemaDef.documentTypeDefMap)
    .map((docDef) => docDef.name)
    .join(', ')

  return `\
// ${autogeneratedNote}

import { ${documentTypeNames}, DocumentTypes } from '../types'

${dataConsts}

export declare const allDocuments: DocumentTypes[]

`
}

const getDataVariableName = ({ docDef }: { docDef: DocumentTypeDef }): string => {
  if (docDef.isSingleton) {
    return utils.lowercaseFirstChar(utils.inflection.singularize(docDef.name))
  } else {
    return 'all' + utils.uppercaseFirstChar(utils.inflection.pluralize(docDef.name))
  }
}

const idToFileName = (id: string): string => {
  return leftPadWithUnderscoreIfStartsWithNumber(id).replace(/\//g, '__')
}

const leftPadWithUnderscoreIfStartsWithNumber = (str: string): string => {
  if (/^[0-9]/.test(str)) {
    return '_' + str
  }
  return str
}

const errorIfArtifactsDirIsDeleted = ({ artifactsDir }: { artifactsDir: string }) => {
  watch(artifactsDir, async (event) => {
    if (event === 'rename' && !(await fileOrDirExists(artifactsDir))) {
      console.error(`Seems like the target directory (${artifactsDir}) was deleted. Please restart the command.`)
      process.exit(1)
    }
  })
}
