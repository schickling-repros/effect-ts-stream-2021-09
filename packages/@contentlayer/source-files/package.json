{
  "name": "@contentlayer/source-files",
  "version": "0.0.22",
  "main": "./dist/cjs/index.js",
  "module": "./dist/esm/index.js",
  "exports": {
    ".": {
      "import": "./dist/esm/index.js",
      "require": "./dist/cjs/index.js"
    },
    "./schema": {
      "import": "./dist/esm/schema/index.js",
      "require": "./dist/cjs/schema/index.js"
    }
  },
  "types": "./dist/esm/index.d.ts",
  "typesVersions": {
    "*": {
      "schema": [
        "./dist/esm/schema"
      ]
    }
  },
  "scripts": {
    "_prepack": "yarn build",
    "dev": "rm -rf dist && tsc --watch",
    "build": "run g:build && echo Build complete"
  },
  "dependencies": {
    "@contentlayer/core": "workspace:*",
    "@contentlayer/utils": "workspace:*",
    "@effect-ts/core": "^0.43.7",
    "@effect-ts/otel": "^0.5.0",
    "chokidar": "^3.5.2",
    "glob": "^7.1.7",
    "glob-promise": "^4.1.0",
    "gray-matter": "^4.0.3",
    "js-yaml": "^4.1.0",
    "minimatch": "^3.0.4",
    "rxjs": "^7.1.0",
    "ts-pattern": "^3.2.1",
    "unified": "^9.2.1"
  },
  "devDependencies": {
    "@types/js-yaml": "^4.0.1",
    "@types/minimatch": "^3.0.4",
    "@types/node": "^15.12.0"
  }
}
