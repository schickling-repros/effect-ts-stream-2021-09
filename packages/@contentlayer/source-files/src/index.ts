import * as core from '@contentlayer/core'
import * as utils from '@contentlayer/utils'
import { FSWatch } from '@contentlayer/utils/node'
import { pipe } from '@effect-ts/core'
import * as T from '@effect-ts/core/Effect'
import * as S from '@effect-ts/core/Effect/Experimental/Stream'
import * as chokidar from 'chokidar'
import type { Observable } from 'rxjs'
import { defer, fromEvent, of } from 'rxjs'
import { map, mergeMap, startWith, tap } from 'rxjs/operators'

import { fetchAllDocuments, makeCacheItemFromFilePath } from './fetchData'
import { makeCoreSchema } from './provideSchema'
import type { DocumentType, Thunk } from './schema'
import type { FilePathPatternMap, PluginOptions } from './types'

export * from './types'
export * from './schema'

type Args = {
  documentTypes: DocumentType[] | Record<string, DocumentType>
  /**
   * Path to the root directory that contains all content. Every content file path will be relative
   * to this directory. This includes:
   *  - `filePathPattern` is relative to `contentDirPath`
   *  - `_raw` fields such as `flattenedPath`, `sourceFilePath`, `sourceFileDir`
   */
  contentDirPath: string
  extensions?: {
    stackbit?: core.StackbitExtension.Config
  }
} & PluginOptions &
  Partial<Flags>

export type Flags = {
  /**
   * Whether to print warning meassages if content has fields not definied in the schema
   * @default 'warn'
   */
  onExtraData: 'warn' | 'ignore'
  /**
   * Whether to skip or fail when encountering missing or incompatible data
   */
  onMissingOrIncompatibleData: 'skip' | 'fail' | 'skip-ignore'
}

type MakeSourcePlugin = (_: Args | Thunk<Args> | Thunk<Promise<Args>>) => Promise<core.SourcePlugin>

export const makeSource: MakeSourcePlugin = async (argsOrArgsThunk) => {
  const {
    contentDirPath,
    documentTypes,
    onMissingOrIncompatibleData = 'skip',
    onExtraData = 'warn',
    extensions,
    ...pluginOptions
  } = typeof argsOrArgsThunk === 'function' ? await argsOrArgsThunk() : argsOrArgsThunk
  const documentTypeDefs = (Array.isArray(documentTypes) ? documentTypes : Object.values(documentTypes)).map((_) =>
    _.def(),
  )
  const schemaDef = { documentTypeDefs }

  const options = {
    markdown: pluginOptions.markdown,
    mdx: pluginOptions.mdx,
    fieldOptions: {
      bodyFieldName: pluginOptions.fieldOptions?.bodyFieldName ?? 'body',
      typeFieldName: pluginOptions.fieldOptions?.typeFieldName ?? 'type',
    },
  }

  const fetchDataEff: any = ({ watch }: any) => {
    const filePathPatternMap: FilePathPatternMap = Object.fromEntries(
      documentTypeDefs
        .filter((_) => _.filePathPattern)
        .map((documentDef) => [documentDef.filePathPattern, documentDef.name]),
    )
    const flags: Flags = { onExtraData, onMissingOrIncompatibleData }

    const coreSchemaDef = makeCoreSchema({ schemaDef, options })
    const initEvent: CustomUpdateEventInit = { _tag: 'init' }

    let cache: core.Cache | undefined = undefined

    const updateStream = pipe(
      FSWatch.makeAndSubscribe('.', {
        cwd: contentDirPath,
        ignoreInitial: true,
        // Unfortunately needed in order to avoid race conditions
        awaitWriteFinish: { stabilityThreshold: 50, pollInterval: 10 },
      }),
      S.map(chokidarAllEventToCustomUpdateEvent_),
      S.tap((e) =>
        T.succeedWith(
          () =>
            (e._tag === 'update' || e._tag === 'deleted') &&
            console.log(`Watch event "${e._tag}": ${e.relativeFilePath}`),
        ),
      ),
      utils.effectUtils.streamStartWith(initEvent),
    )

    return pipe(
      updateStream,
      S.tap(() =>
        T.tryCatchPromise(
          async () => {
            if (!cache) {
              cache = await core.loadPreviousCacheFromDisk({ schemaHash: coreSchemaDef.hash })
            }
          },
          (e) => e as Error,
        ),
      ),
      S.mapEff((event) =>
        T.tryCatchPromise(
          async () => {
            switch (event._tag) {
              case 'init':
                return fetchAllDocuments({
                  coreSchemaDef,
                  filePathPatternMap,
                  contentDirPath,
                  flags,
                  options,
                  previousCache: cache,
                })
              case 'deleted': {
                delete cache!.cacheItemsMap[event.relativeFilePath]
                return cache!
              }
              case 'update': {
                return updateCacheEntry({
                  contentDirPath,
                  filePathPatternMap,
                  cache: cache!,
                  event,
                  flags,
                  coreSchemaDef,
                  options,
                })
              }
              default:
                utils.casesHandled(event)
            }
          },
          (e) => e as Error,
        ),
      ),
      S.take(1),
      S.runDrain,
    )
  }

  const fetchData: core.FetchDataFn = ({ watch }) => {
    const filePathPatternMap: FilePathPatternMap = Object.fromEntries(
      documentTypeDefs
        .filter((_) => _.filePathPattern)
        .map((documentDef) => [documentDef.filePathPattern, documentDef.name]),
    )
    const flags: Flags = { onExtraData, onMissingOrIncompatibleData }

    const coreSchemaDef = makeCoreSchema({ schemaDef, options })
    const initEvent: CustomUpdateEventInit = { _tag: 'init' }

    let cache: core.Cache | undefined = undefined

    const updates$: Observable<CustomUpdateEvent> = watch
      ? defer(
          () =>
            fromEvent(
              chokidar.watch('.', {
                cwd: contentDirPath,
                ignoreInitial: true,
                // Unfortunately needed in order to avoid race conditions
                awaitWriteFinish: { stabilityThreshold: 50, pollInterval: 10 },
              }),
              'all',
            ) as Observable<ChokidarAllEvent>,
        ).pipe(
          map(chokidarAllEventToCustomUpdateEvent),
          tap(
            (e) =>
              (e._tag === 'update' || e._tag === 'deleted') &&
              console.log(`Watch event "${e._tag}": ${e.relativeFilePath}`),
          ),
          startWith(initEvent),
        )
      : of(initEvent)

    return (
      updates$
        .pipe(
          // init cache from persisted cache
          mergeMap(async (event) => {
            if (!cache) {
              cache = await core.loadPreviousCacheFromDisk({ schemaHash: coreSchemaDef.hash })
            }
            return event
          }),
          mergeMap(async (event) => {
            switch (event._tag) {
              case 'init':
                return fetchAllDocuments({
                  coreSchemaDef,
                  filePathPatternMap,
                  contentDirPath,
                  flags,
                  options,
                  previousCache: cache,
                })
              case 'deleted': {
                delete cache!.cacheItemsMap[event.relativeFilePath]
                return cache!
              }
              case 'update': {
                return updateCacheEntry({
                  contentDirPath,
                  filePathPatternMap,
                  cache: cache!,
                  event,
                  flags,
                  coreSchemaDef,
                  options,
                })
              }
              default:
                utils.casesHandled(event)
            }
          }),
        )
        // update local and persisted cache
        .pipe(
          tap(async (cache_) => {
            cache = cache_
            await core.writeCacheToDisk({ cache, schemaHash: coreSchemaDef.hash })
          }),
        )
    )
  }

  return {
    type: 'local',
    extensions: extensions ?? {},
    options,
    provideSchema: () => makeCoreSchema({ schemaDef, options }),
    provideSchemaEff: T.succeedWith(() => makeCoreSchema({ schemaDef, options })),
    fetchData,
    fetchDataEff,
  }
}

const updateCacheEntry = async ({
  contentDirPath,
  filePathPatternMap,
  cache,
  event,
  flags,
  coreSchemaDef,
  options,
}: {
  contentDirPath: string
  filePathPatternMap: FilePathPatternMap
  cache: core.Cache
  event: CustomUpdateEventFileUpdated
  flags: Flags
  coreSchemaDef: core.SchemaDef
  options: core.PluginOptions
}): Promise<core.Cache> => {
  const cacheItem = await makeCacheItemFromFilePath({
    relativeFilePath: event.relativeFilePath,
    contentDirPath,
    filePathPatternMap,
    flags,
    coreSchemaDef,
    options,
    previousCache: cache,
  })

  if (cacheItem) {
    cache!.cacheItemsMap[event.relativeFilePath] = cacheItem
  }

  return cache
}

const chokidarAllEventToCustomUpdateEvent = ([eventName, relativeFilePath]: ChokidarAllEvent): CustomUpdateEvent => {
  switch (eventName) {
    case 'add':
    case 'change':
      return { _tag: 'update', relativeFilePath }
    case 'unlink':
      return { _tag: 'deleted', relativeFilePath }
    case 'unlinkDir':
    case 'addDir':
      return { _tag: 'init' }
    default:
      utils.casesHandled(eventName)
  }
}

const chokidarAllEventToCustomUpdateEvent_ = (event: FSWatch.FileSystemEvent): CustomUpdateEvent => {
  switch (event._tag) {
    case 'FileAdded':
    case 'FileChanged':
      return { _tag: 'update', relativeFilePath: event.path }
    case 'FileRemoved':
      return { _tag: 'deleted', relativeFilePath: event.path }
    case 'DirectoryRemoved':
    case 'DirectoryAdded':
      return { _tag: 'init' }
    default:
      utils.casesHandled(event)
  }
}

type ChokidarAllEvent = [eventName: 'add' | 'addDir' | 'change' | 'unlink' | 'unlinkDir', path: string, stats?: any]

type CustomUpdateEvent = CustomUpdateEventFileUpdated | CustomUpdateEventFileDeleted | CustomUpdateEventInit

type CustomUpdateEventFileUpdated = {
  readonly _tag: 'update'
  relativeFilePath: string
}

type CustomUpdateEventFileDeleted = {
  readonly _tag: 'deleted'
  relativeFilePath: string
}

type CustomUpdateEventInit = {
  readonly _tag: 'init'
}
