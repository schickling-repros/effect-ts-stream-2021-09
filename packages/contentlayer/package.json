{
  "name": "contentlayer",
  "version": "0.0.22",
  "bin": "./dist/cjs/cli/index.js",
  "exports": {
    "./source-files": {
      "import": "./dist/esm/source-files/index.js",
      "require": "./dist/cjs/source-files/index.js"
    },
    "./source-files/schema": {
      "import": "./dist/esm/source-files/schema/index.js",
      "require": "./dist/cjs/source-files/schema/index.js"
    },
    "./client": {
      "import": "./dist/esm/client/index.js",
      "require": "./dist/cjs/client/index.js"
    },
    "./utils": {
      "import": "./dist/esm/utils/index.js",
      "require": "./dist/cjs/utils/index.js"
    },
    "./utils/node": {
      "import": "./dist/esm/utils/node/index.js",
      "require": "./dist/cjs/utils/node/index.js"
    },
    "./core": {
      "import": "./dist/esm/core/index.js",
      "require": "./dist/cjs/core/index.js"
    }
  },
  "_typesVersions": "Not needed anymore once this issue is closed https://github.com/microsoft/TypeScript/issues/33079",
  "typesVersions": {
    "*": {
      "source-files": [
        "./dist/esm/source-files"
      ],
      "source-files/schema": [
        "./dist/esm/source-files/schema"
      ],
      "client": [
        "./dist/esm/client"
      ],
      "utils": [
        "./dist/esm/utils"
      ],
      "utils/node": [
        "./dist/esm/utils/node"
      ],
      "core": [
        "./dist/esm/core"
      ]
    }
  },
  "scripts": {
    "_prepack": "yarn build",
    "build": "run g:build && echo Build complete",
    "_build:cli": "rm -rf dist/cli && esbuild --bundle ./src/cli/index.ts --outfile=./dist/cli/index.js --platform=node --target=node12 --sourcemap --external:esbuild && chmod +x ./dist/cli/index.js"
  },
  "dependencies": {
    "@contentlayer/cli": "workspace:*",
    "@contentlayer/client": "workspace:*",
    "@contentlayer/core": "workspace:*",
    "@contentlayer/source-files": "workspace:*",
    "@contentlayer/utils": "workspace:*"
  },
  "devDependencies": {
    "esbuild": "^0.12.14",
    "typescript": "^4.4.2"
  }
}
