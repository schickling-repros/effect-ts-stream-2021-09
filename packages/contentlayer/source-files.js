module.exports = require('./dist/cjs/source-files')
