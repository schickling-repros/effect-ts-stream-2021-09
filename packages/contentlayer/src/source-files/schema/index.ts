export * from '@contentlayer/source-files/schema'
