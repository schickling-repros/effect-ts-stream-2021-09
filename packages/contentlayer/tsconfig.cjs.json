{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "module": "commonjs",
    "outDir": "./dist/cjs",
    "tsBuildInfoFile": "./dist/.tsbuildinfo.cjs.json"
  },
  "references": [
    { "path": "../@contentlayer/cli/tsconfig.cjs.json" },
    { "path": "../@contentlayer/client/tsconfig.cjs.json" },
    { "path": "../@contentlayer/source-files/tsconfig.cjs.json" },
    { "path": "../@contentlayer/utils/tsconfig.cjs.json" },
    { "path": "../@contentlayer/core/tsconfig.cjs.json" }
  ]
}
