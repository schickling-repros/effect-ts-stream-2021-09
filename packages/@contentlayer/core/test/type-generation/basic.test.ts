import { defineDocumentType, makeSource } from 'contentlayer/source-files'

import { renderTypes } from '../../src/generation/generate-types'

test.todo('generate-types')

const TestPost = defineDocumentType<any>(() => ({
  name: 'TestPost',
  filePathPattern: `**/*.md`,
  fields: {
    title: {
      type: 'string',
      description: 'The title of the post',
      required: true,
    },
    date: {
      type: 'date',
      description: 'The date of the post',
      required: true,
    },
  },
  computedFields: {
    slug: { type: 'string', resolve: (_) => _._id.replace('.md', '') },
  },
}))

// TODO rewrite test for gendotpkg
describe('generate-types', () => {
  test('simple schema', async () => {
    const sourcePlugin = await makeSource({ documentTypes: [TestPost], contentDirPath: '' })
    const schemaDef = await sourcePlugin.provideSchema()
    const typeSource = renderTypes({
      schemaDef,
      generationOptions: {
        sourcePluginType: 'local',
        options: {
          fieldOptions: { bodyFieldName: 'body', typeFieldName: 'type' },
          markdown: undefined,
          mdx: undefined,
        },
      },
    })
    expect(typeSource).toMatchSnapshot()
  })
})
