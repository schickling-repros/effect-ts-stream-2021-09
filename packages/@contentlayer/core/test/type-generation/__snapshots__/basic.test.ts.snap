// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate-types simple schema 1`] = `
"// NOTE This file is auto-generated by the Contentlayer CLI

import type { Markdown, MDX } from 'contentlayer/core'
import * as Local from 'contentlayer/source-files'

export { isType } from 'contentlayer/client'

// export type Image = string
export type { Markdown, MDX }

export interface ContentlayerGenTypes {
  documentTypes: DocumentTypes
  documentTypeMap: DocumentTypeMap
  documentTypeNames: DocumentTypeNames
  nestedTypes: NestedTypes
  nestedTypeMap: NestedTypeMap
  nestedTypeNames: NestedTypeNames
  allTypeNames: AllTypeNames
}

declare global {
  interface ContentlayerGen extends ContentlayerGenTypes {}
}

export type DocumentTypeMap = {
  TestPost: TestPost
}

export type NestedTypeMap = {

}

export type AllTypes = DocumentTypes | NestedTypes
export type AllTypeNames = DocumentTypeNames | NestedTypeNames

export type DocumentTypes = TestPost
export type DocumentTypeNames = DocumentTypes['_typeName']

export type NestedTypes = never
export type NestedTypeNames = NestedTypes['_typeName']



/** Document types */
export type TestPost = {
  /** File path relative to \`contentDirPath\` */
  _id: string
  _typeName: 'TestPost'
  _raw: Local.RawDocumentData
  /** The title of the post */
  title: string
  /** The date of the post */
  date: string
  /** Markdown file body */
  body: Markdown
  slug: string
}  

/** Nested types */
  
  
 "
`;
